notes/trees.md

```diff
@@ -0,0 +1,234 @@
+# Trees.scala
+
+## Top level (`meta.Source`, `meta.Pkg`, `meta.Pkg`)
+
+```scala
+"package a".parse[Source].get.structure
+```
+
+## Declarations (`meta.Decl`)
+
+```scala
+"val a: Int".parse[Stat].get.structure
+```
+
+Decl            | Examples
+----------------|-----------
+Decl.Val        | val a
+Decl.Var        | var a
+Decl.Def        | def f: Int
+Decl.Type       | type T
+
+
+## Definitions (`meta.Defn`)
+
+```scala
+"val a = 1".parse[Stat].get.structure
+```
+
+Defn        | Examples
+------------|-----------
+Defn.Val    | val a = 1
+Defn.Var    | var a = 1
+Defn.Def    | def f = 1
+Defn.Macro  | def f = macro impl
+Defn.Type   | type T = Int
+Defn.Class  | class A
+Defn.Trait  | trait A
+Defn.Object | object A
+
+## Terms (`meta.Term`)
+
+```scala
+"1 + 1".parse[Term].get.structure
+```
+
+Ref             | Examples
+----------------|-----------
+Term.This       | `this`, `foo.this`
+Term.Super      | `super`, `foo.super`, `super[bar]"`, `foo.super[bar]`
+Term.Name       | `foo`
+Term.Select     | `bar.foo`
+Term.ApplyUnary | `!foo`, `~foo`, `-foo`, `foo`
+
+Term                  | Examples
+----------------------|------------------------------
+Term.Apply            | `buzz(42)`
+Term.ApplyType        | `implicitly[Ordering[Int]]`
+Term.ApplyInfix       | `a + a`
+Term.Assign           | `a = 1`
+Term.Return           | `return bar`
+Term.Throw            | `throw exception`
+Term.Ascribe          | `a: Int`
+Term.Annotate         | `(x: @unchecked)`
+Term.Tuple            | `(1, 1)`
+Term.Block            | `{ bar; buzz }`
+Term.If               | `if (p) t else f`
+Term.Match            | `x match { case _ => }`
+Term.Try              | `try foo catch { case _ => }`
+Term.TryWithHandler   | `try foo catch (h)`
+Term.Function         | `(a, b) => a + b`
+Term.PartialFunction  | `{ case _ =>  }`
+Term.While            | `while(p){ d }`
+Term.Do               | `do { d } while(p)`
+Term.For              | `for{x <- xs} f(x)`
+Term.ForYield         | `for{x <- xs} yield f(x)`
+Term.New              | `new A`
+Term.NewAnonymous     | `new A { }`
+Term.Placeholder      | `_`
+Term.Eta              | `foo _`
+Term.Repeated         | `foo(x: _*)`
+Term.Param            | `def foo(x: Int)`
+Term.Interpolate      | `s"foo ~ $bar ~ ${42}"`
+Term.Xml              | `<h1>Hello ~ {bar} ~ {42}</h1>`
+
+## Types (`meta.Type`)
+
+```scala
+"A with B".parse[Type].get.structure
+```
+
+Ref            | Examples
+--------------------------
+Type.Name      | `B`
+Type.Select    | `a.B`
+Type.Project   | `a#B`
+Type.Singleton | `t.type`
+
+Type                  | Examples
+----------------------|--------------------
+Type.Apply            | `F[T]`
+Type.ApplyInfix       | `K Map V`
+Type.Function         | `A => B`
+Type.ImplicitFunction | `implicit A => B`
+Type.Tuple            | `(A, B)`
+Type.With             | `A with B`
+Type.And              | `A & B`
+Type.Or               | `A \| B`
+Type.Refine           | `A { def f: Int }`
```

Comment:
`T { def f1: Int; def f2: Int }`

Reply:
Why two stats ?

Reply:
It can be a single stat `def f: Int`. It we want to be consistent with `Term.Block` however, we should update every place where a list of  arguments is expected to use `f1(); f2(); ...`. Overkill?